Fix product shortcode pagination

claudiulodro opened the pull request:
Fixes the FINAL TODO mentioned in #17683

claudiulodro committed:
Fix product shortcode pagination

claudiulodro committed:
line up array

claudiulodro committed:
Fix weird bracket

mikejolley commented on includes/wc-template-functions.php:
```diff
@@ -895,12 +893,10 @@ function woocommerce_result_count( $query = null ) {
-			$query = $GLOBALS['wp_query'];
-		}
+	function woocommerce_catalog_ordering( $args = array() ) {
+		$query = $GLOBALS['wp_query'];
```
I'd switch the logic up here somewhat and in the other function so it's more readable.

- If $args is empty, inherit props from global query object.
- Include array of default args and use wp_parse_args
- Within the function, you can then use the variable names after parsing instead of checking isset multiple times. This also let's us whitelist what is used.

mikejolley replied:
Also let's you change the variable naming if needed.

mikejolley commented on includes/wc-template-functions.php:
```diff
@@ -942,16 +938,14 @@ function woocommerce_catalog_ordering( $query = null ) {
-			$query = $GLOBALS['wp_query'];
-		}
+	function woocommerce_pagination( $args = array() ) {
+		$query = $GLOBALS['wp_query'];
```
See comment under woocommerce_catalog_ordering

mikejolley commented on includes/shortcodes/class-wc-shortcode-products.php:
```diff
@@ -504,10 +505,18 @@ protected function get_products_ids() {
 			}
-			$ids = wp_parse_id_list( $products->posts );
+
+			$ids = array(
```
Maybe name this $return now it's not just ids? Nicpick.

claudiosanches replied:
Or just `$data` shoulds better.

mikejolley commented on includes/shortcodes/class-wc-shortcode-products.php:
```diff
@@ -534,10 +543,10 @@ protected function product_loop() {
 		ob_start();
 
-		if ( $products_ids ) {
+		if ( $products_ids && $products_ids['ids'] ) {
```
Since we are changing what `get_products_ids` returns, do we need a new method?

claudiulodro replied:
I don't believe so. It's protected and it's only been around since 3.2.4. Before that it was called `get_products`. The only place it's used in in `product_loop` below.

claudiosanches replied:
I prefer we change the name of the method again.

mikejolley reviewed (changes requested):
Added feedback.

Did have an alternative idea though - just needs testing to see how much of a negative performance impact it would have...

If we stuck with queries (vs args), and cached IDs, could we pass the cached IDs into a new WP_Query object with `post__in`? If thats not slower than we have now we can get the best of both worlds.

claudiulodro committed:
Refactor for better style

claudiulodro committed:
Line up arrays

claudiulodro commented:
Good feedback. This should be ready to go.
- Refactored template functions to make them cleaner
- Renamed shortcode methods/variables to make them more correct
- Fixes phpcs stuff in wc-template-functions.php

mikejolley committed:
Normalise get_query_results names and add new param for pagination

claudiosanches reviewed (approved):
LGTM
